Java API for Pluggable components can't run on MacOS

## Describe the feature

While trying to run the pluggable components Java SDK on MacOS users receive the following error:


```
INFO: Taking unix socket domain path from env. var. DAPR_COMPONENT_SOCKET_PATH
Sep 22, 2022 4:35:54 PM org.example.StateStoreComponentServer main
INFO: Starting a StateStoreComponentServer...
Sep 22, 2022 4:35:54 PM org.example.StateStoreComponentServer buildUnixSocketServer
INFO: Configuring server to listen to unix socket domain on file /tmp/sharedUDS/javaMemstore.socket
Exception in thread "main" java.lang.UnsatisfiedLinkError: failed to load the required native library
    at io.netty.channel.epoll.Epoll.ensureAvailability(Epoll.java:81)
    at io.netty.channel.epoll.EpollEventLoop.<clinit>(EpollEventLoop.java:54)
    at io.netty.channel.epoll.EpollEventLoopGroup.newChild(EpollEventLoopGroup.java:186)
    at io.netty.channel.epoll.EpollEventLoopGroup.newChild(EpollEventLoopGroup.java:37)
    at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:84)
    at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:60)
    at io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:49)
    at io.netty.channel.MultithreadEventLoopGroup.<init>(MultithreadEventLoopGroup.java:59)
    at io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:114)
    at io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:101)
    at io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:78)
    at io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:54)
    at io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:47)
    at org.example.StateStoreComponentServer.buildUnixSocketServer(StateStoreComponentServer.java:86)
    at org.example.StateStoreComponentServer.main(StateStoreComponentServer.java:63)
Caused by: java.lang.ExceptionInInitializerError
    at io.netty.channel.epoll.Epoll.<clinit>(Epoll.java:40)
    ... 14 more
Caused by: java.lang.IllegalStateException: Only supported on Linux
    at io.netty.channel.epoll.Native.loadNativeLibrary(Native.java:317)
    at io.netty.channel.epoll.Native.<clinit>(Native.java:85)
    ... 15 more

```

This might be a problem with Netty support for UDS but regardless it needs to be addressed. FWIW, this might be the wrong repo for this issue.

--- a/components-java-sdk-examples/pom.xml
+++ b/components-java-sdk-examples/pom.xml
@@ -1,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.dapr</groupId>
+        <artifactId>components-java-sdk-parent</artifactId>
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>components-java-sdk-examples</artifactId>
+    <packaging>jar</packaging>
+    <version>1.0-SNAPSHOT</version>
+    <name>components-java-sdk-examples</name>
+    <description>Examples for Dapr Java Pluggable Components SDK</description>
+
+    <properties>
+        <maven.compiler.source>17</maven.compiler.source>
+        <maven.compiler.target>17</maven.compiler.target>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.dapr</groupId>
+            <artifactId>components-java-sdk</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+    </dependencies>
+
+</project>

--- a/components-java-sdk-examples/src/main/java/io/dapr/examples/InMemoryStateStorePluggableComponentServer.java
+++ b/components-java-sdk-examples/src/main/java/io/dapr/examples/InMemoryStateStorePluggableComponentServer.java
@@ -1,0 +1,39 @@
+/*
+ * Copyright 2023 The Dapr Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dapr.examples;
+
+import io.dapr.components.server.PluggableComponent;
+import io.dapr.components.server.PluggableComponentServer;
+import io.dapr.examples.statestore.InMemoryStateStore;
+
+import java.io.IOException;
+
+public class InMemoryStateStorePluggableComponentServer {
+
+  /**
+   * Main body for our example.
+   *
+   * @param args Command line arguments that we ignore.
+   * @throws IOException if we face errors starting this server.
+   */
+  public static void main(String[] args) throws IOException {
+
+    final PluggableComponentServer server = new PluggableComponentServer();
+    server.registerComponent(PluggableComponent
+            .withName("my-in-memory-state-store")
+            .withStateStore(new InMemoryStateStore())
+        )
+        .run();
+  }
+}

--- a/components-java-sdk-examples/src/main/java/io/dapr/examples/InMemoryStateStorePluggableComponentServer.java
+++ b/components-java-sdk-examples/src/main/java/io/dapr/examples/InMemoryStateStorePluggableComponentServer.java
@@ -34,6 +34,10 @@ public class InMemoryStateStorePluggableComponentServer {
             .withName("my-in-memory-state-store")
             .withStateStore(new InMemoryStateStore())
         )
+        .registerComponent(PluggableComponent
+            .withName("another-in-memory-state-store")
+            .withStateStore(new InMemoryStateStore())
+        )
         .run();
   }
 }

--- a/components-java-sdk-examples/src/main/java/io/dapr/examples/statestore/InMemoryStateStore.java
+++ b/components-java-sdk-examples/src/main/java/io/dapr/examples/statestore/InMemoryStateStore.java
@@ -31,14 +31,8 @@ public class InMemoryStateStore implements StateStore {
 
   @Override
   public Mono<Void> init(Map<String, String> properties) {
-    // log.info("Initializing InMemoryState Store."); // I was born ready, babe
     return Mono.empty();
   }
-
-  //  @Override
-  //  public Mono<List<String>> getFeatures() {
-  //    return Mono.just(Collections.emptyList());
-  //  }
 
   @Override
   public Mono<GetResponse> get(GetRequest getRequest) {

--- a/components-java-sdk-examples/src/main/java/io/dapr/examples/statestore/InMemoryStateStore.java
+++ b/components-java-sdk-examples/src/main/java/io/dapr/examples/statestore/InMemoryStateStore.java
@@ -1,0 +1,77 @@
+/*
+ * Copyright 2022 The Dapr Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dapr.examples.statestore;
+
+
+import io.dapr.components.domain.state.DeleteRequest;
+import io.dapr.components.domain.state.Exceptions;
+import io.dapr.components.domain.state.GetRequest;
+import io.dapr.components.domain.state.GetResponse;
+import io.dapr.components.domain.state.SetRequest;
+import io.dapr.components.domain.state.StateStore;
+import reactor.core.publisher.Mono;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class InMemoryStateStore implements StateStore {
+
+  private final Map<String, GetResponse> entries = new HashMap<>();
+
+  @Override
+  public Mono<Void> init(Map<String, String> properties) {
+    // log.info("Initializing InMemoryState Store."); // I was born ready, babe
+    return Mono.empty();
+  }
+
+  //  @Override
+  //  public Mono<List<String>> getFeatures() {
+  //    return Mono.just(Collections.emptyList());
+  //  }
+
+  @Override
+  public Mono<GetResponse> get(GetRequest getRequest) {
+    return Mono.justOrEmpty(entries.get(getRequest.key()));
+  }
+
+  @Override
+  public Mono<Void> delete(DeleteRequest deleteRequest) {
+    final String key = deleteRequest.key();
+    final GetResponse existingData = entries.get(key);
+    if (existingData == null) {
+      // Key not found: nothing to remove, nothing else to do.
+      return Mono.empty();
+    } else if (existingData.etag().equals(deleteRequest.etag())) {
+      // Valid remove request
+      entries.remove(key);
+      return Mono.empty();
+    } else {
+      // Data exists but etag mismatch
+      return Mono.error(Exceptions.getEtagMismatchException("in-memory-etag-mismatch-on-delete"));
+    }
+  }
+
+  @Override
+  public Mono<Void> set(SetRequest setRequest) {
+    var upsertEntry = new GetResponse(
+        setRequest.value(),
+        setRequest.etag(),
+        setRequest.metadata(),
+        setRequest.contentType()
+    );
+    entries.put(setRequest.key(), upsertEntry);
+    // Nothing else to do but return with no errors.
+    return Mono.empty();
+  }
+}

--- a/components-java-sdk/pom.xml
+++ b/components-java-sdk/pom.xml
@@ -39,25 +39,25 @@
         <dependency>
             <groupId>io.netty</groupId>
             <artifactId>netty-transport-native-kqueue</artifactId>
-            <version>4.1.79.Final</version>
+            <version>${netty-transport-native.version}</version>
             <classifier>osx-aarch_64</classifier>
         </dependency>
         <dependency>
             <groupId>io.netty</groupId>
             <artifactId>netty-transport-native-kqueue</artifactId>
-            <version>4.1.79.Final</version>
+            <version>${netty-transport-native.version}</version>
             <classifier>osx-x86_64</classifier>
         </dependency>
         <dependency>
             <groupId>io.netty</groupId>
             <artifactId>netty-transport-native-epoll</artifactId>
-            <version>4.1.79.Final</version>
+            <version>${netty-transport-native.version}</version>
             <classifier>linux-x86_64</classifier>
         </dependency>
         <dependency>
             <groupId>io.netty</groupId>
             <artifactId>netty-transport-native-epoll</artifactId>
-            <version>4.1.79.Final</version>
+            <version>${netty-transport-native.version}</version>
             <classifier>linux-aarch_64</classifier>
         </dependency>
     </dependencies>

--- a/components-java-sdk/pom.xml
+++ b/components-java-sdk/pom.xml
@@ -27,11 +27,38 @@
             <artifactId>components-java-sdk-autogen</artifactId>
             <version>${project.version}</version>
         </dependency>
+
+        <!--This is our async framework. Dapr Java SDK uses the same -->
         <dependency>
             <groupId>io.projectreactor</groupId>
             <artifactId>reactor-core</artifactId>
             <version>3.3.11.RELEASE</version>
         </dependency>
-    </dependencies>
 
+        <!-- Netty's Unix Domain Socket support for Linux and Mac, for both AMD64 and ARM64 -->
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-transport-native-kqueue</artifactId>
+            <version>4.1.79.Final</version>
+            <classifier>osx-aarch_64</classifier>
+        </dependency>
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-transport-native-kqueue</artifactId>
+            <version>4.1.79.Final</version>
+            <classifier>osx-x86_64</classifier>
+        </dependency>
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-transport-native-epoll</artifactId>
+            <version>4.1.79.Final</version>
+            <classifier>linux-x86_64</classifier>
+        </dependency>
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-transport-native-epoll</artifactId>
+            <version>4.1.79.Final</version>
+            <classifier>linux-aarch_64</classifier>
+        </dependency>
+    </dependencies>
 </project>

--- a/components-java-sdk/src/main/java/io/dapr/components/domain/state/Exceptions.java
+++ b/components-java-sdk/src/main/java/io/dapr/components/domain/state/Exceptions.java
@@ -1,0 +1,121 @@
+/*
+ * Copyright 2023 The Dapr Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dapr.components.domain.state;
+
+import com.google.protobuf.Any;
+import com.google.rpc.BadRequest;
+import com.google.rpc.ErrorInfo;
+import com.google.rpc.Status;
+import io.grpc.protobuf.StatusProto;
+
+import java.util.Map;
+
+/**
+ * Utility functions to wrap gRPC error Status into exceptions.
+ *
+ * <p>You might use these functions with {@code Mono.error} and {@code Flux.error} methods.</p>
+ */
+public class Exceptions {
+
+  private static final String AFFECTED = "affected";
+  private static final String EXPECTED = "expected";
+  private static final String ETAG_FIELD_NAME = "etag";
+
+  /**
+   * The default description for Etag Mismatch errors.
+   */
+  public static final String ETAG_MISMATCH_DESCRIPTION = "The ETag field provided does not match the one in the store";
+
+  /**
+   * Builds an exception wrapping an gRPC status error for an Etag Mismatch Error.
+   *
+   * @param message a message-code for this error.
+   * @param description A message describing how/why this etag mismatched
+   * @return A gRPC compatible exception.
+   */
+  public static Throwable getEtagMismatchException(final String message, final String description) {
+    final Status status = Status.newBuilder()
+        .setCode(io.grpc.Status.Code.FAILED_PRECONDITION.value())
+        .setMessage(message)
+        .addDetails(Any.pack(BadRequest.FieldViolation.newBuilder()
+            .setField(ETAG_FIELD_NAME)
+            .setDescription(description)
+            .build()))
+        .build();
+    return StatusProto.toStatusException(status);
+  }
+
+  /**
+   * Builds an exception wrapping an gRPC status error for an Etag Mismatch Error with
+   * a default description.
+   *
+   * @param message a message-code for this error.
+   * @return A gRPC compatible exception.
+   */
+  public static Throwable getEtagMismatchException(final String message) {
+    final Status status = Status.newBuilder()
+        .setCode(io.grpc.Status.Code.FAILED_PRECONDITION.value())
+        .setMessage(message)
+        .addDetails(Any.pack(BadRequest.FieldViolation.newBuilder()
+            .setField(ETAG_FIELD_NAME)
+            .setDescription(ETAG_MISMATCH_DESCRIPTION)
+            .build()))
+        .build();
+    return StatusProto.toStatusException(status);
+  }
+
+  /**
+   * Builds an exception wrapping an gRPC status error for an Invalid Etag.
+   *
+   * @param message a message-code for this error.
+   * @param description A message describing why this etag is invalid.
+   * @return A gRPC compatible exception.
+   */
+  public static Throwable getInvalidETagException(final String message, final String description) {
+    final Status status = Status.newBuilder()
+        .setCode(io.grpc.Status.Code.INVALID_ARGUMENT.value())
+        .setMessage(message)
+        .addDetails(Any.pack(BadRequest.FieldViolation.newBuilder()
+            .setField(ETAG_FIELD_NAME)
+            .setDescription(description)
+            .build()))
+        .build();
+    return StatusProto.toStatusException(status);
+  }
+
+  /**
+   * Builds an exception wrapping an gRPC status error when the number of items affected
+   * by a bulk operation does not match the number of requested items in the bulk request.
+   *
+   * @param message a message-code for this error.
+   * @param affected Number of items that were ultimately affected by the bulk operation.
+   * @param expected Number of items that were originally expected to be operated by bulk operation.
+   * @return A gRPC compatible exception.
+   */
+  public static Throwable getBulkDeleteRowMismatchException(final String message,
+                                                            final int affected,
+                                                            final int expected) {
+    final Status status = Status.newBuilder()
+        .setCode(io.grpc.Status.Code.INTERNAL.value())
+        .setMessage(message)
+        .addDetails(Any.pack(ErrorInfo.newBuilder()
+            .putAllMetadata(Map.ofEntries(
+                Map.entry(AFFECTED, Integer.toString(affected)),
+                Map.entry(EXPECTED, Integer.toString(expected))
+            ))
+            .build()))
+        .build();
+    return StatusProto.toStatusException(status);
+  }
+}

--- a/components-java-sdk/src/main/java/io/dapr/components/server/Constants.java
+++ b/components-java-sdk/src/main/java/io/dapr/components/server/Constants.java
@@ -1,0 +1,57 @@
+/*
+ * Copyright 2023 The Dapr Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dapr.components.server;
+
+/**
+ * Constant definitions.
+ */
+public final class Constants {
+  public static final class Defaults {
+    /**
+     * Default location where Dapr looks for Pluggable Components Unix Domain Socket files.
+     * If env. var named by {@link EnvironmentVariable#DAPR_COMPONENTS_SOCKETS_PATH} is not set, this value is used.
+     */
+    public static final String DAPR_COMPONENTS_SOCKETS_PATH = "/tmp/dapr-components-sockets";
+
+    /**
+     * The default suffix for socket files created by Dapr Pluggable Components.
+     *
+     * <p>Can be overridden by  {@link EnvironmentVariable#DAPR_COMPONENTS_SOCKET_EXTENSION}.
+     */
+    public static final String DAPR_COMPONENTS_SOCKET_EXTENSION = "sock";
+
+
+  }
+
+  public static final class EnvironmentVariable {
+
+    /**
+     * Environment Variable specifying where Dapr will look for Pluggable Components Unix Domain Socket files.
+     * If specified, defines were this application will write the UDS files it creates.
+     * If unset, the value from {@link Defaults#DAPR_COMPONENTS_SOCKETS_PATH} is used instead.
+     *
+     */
+    public static final String DAPR_COMPONENTS_SOCKETS_PATH = "DAPR_COMPONENTS_SOCKETS_FOLDER";
+
+    /**
+     * The environment variable name that defines the extension for socket files created by Dapr Pluggable Components.
+     *
+     * <p>If unset, the value from {@link Defaults#DAPR_COMPONENTS_SOCKET_EXTENSION} is used instead.
+     * If set, the unix domain socket file for components will be built by joining a component name,
+     * a single "." character and the contents of this variable.
+     */
+    public static final String DAPR_COMPONENTS_SOCKET_EXTENSION = "DAPR_COMPONENTS_SOCKETS_EXTENSION";
+  }
+
+}

--- a/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponent.java
+++ b/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponent.java
@@ -1,0 +1,108 @@
+/*
+ * Copyright 2023 The Dapr Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dapr.components.server;
+
+import com.google.common.collect.ImmutableList;
+import io.dapr.components.domain.state.StateStore;
+import io.dapr.components.wrappers.StateStoreGrpcComponentWrapper;
+import io.grpc.BindableService;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Represents a single pluggable component.
+ *
+ * <p>A pluggable component can be accessed by a single Unix Domain Socket file.
+ * The name of this component is used to generate this UDS filename.
+ *
+ * <p>Additionally, a single pluggable component can host different components APIs.
+ * No, that is not the normal use-case, but it is a supported one. For further
+ * information see
+ * <a href="https://docs.dapr.io/operations/components/pluggable-components/pluggable-components-registration/">
+ * How-To: Register a pluggable component</a>
+ *
+ * <p>Technically, this class represents a named collection of gRPC <b>services</b> implementing
+ * distinct Components API that will, when materialized by a {@link PluggableComponentServer},
+ * have a 1:1 mapping with a Unix Domain Socket file</p>
+ *
+ * <p>A PluggableComponent instance by itself is not capable of receiving and handling requests:
+ * you need to add it to a {@link PluggableComponentServer}.</p>
+ */
+public final class PluggableComponent {
+  private final String name;
+
+  private final List<BindableService> exposedServices = new ArrayList<>();
+
+  // A given component API (state store, pubusb, binding etc) can be registered only once.
+  private boolean alreadyAddedStaStore = false;
+
+  /**
+   * Creates a new pluggable component.
+   *
+   * @param componentName this pluggable component name.
+   */
+  public PluggableComponent(String componentName) {
+    this.name = Objects.requireNonNull(componentName);
+  }
+
+  /**
+   * Static builder for PluggableComponents.
+   *
+   * @param componentName name for the new pluggable component.
+   * @return a PluggableComponent instance.
+   */
+  public static PluggableComponent withName(String componentName) {
+    return new PluggableComponent(componentName);
+  }
+
+  /**
+   * Register a new StateStore with this pluggable component.
+   *
+   * @param stateStore The stateStore instance we want to expose through this pluggable component.
+   * @return The current {@link PluggableComponent} instance, so calls can be chained.
+   */
+  public PluggableComponent withStateStore(StateStore stateStore) {
+    assert !alreadyAddedStaStore; // No, you cannot add multiple stateStores
+
+    exposedServices.add(new StateStoreGrpcComponentWrapper(stateStore));
+
+    // Register other facets of a stateStore like QueriableStateStore and TransactionalStateStore
+    // IFF the current stateStore object supports those facets.
+    //
+    // if (stateStore instanceof QueriableStateStore) {
+    //   exposedServices.add(new QueriableStateStoreGrpcComponentWrapper(stateStore));
+    // }
+    // if (stateStore instanceof TransactionalStateStore) {
+    //   exposedServices.add(new TransactionalStateStoreGrpcComponentWrapper(stateStore));
+    // }
+
+    // StateStore added and nothing else can be added
+    alreadyAddedStaStore = true;
+    return this;
+  }
+
+  //
+  // Getters are package private
+  //
+
+  String getName() {
+    return name;
+  }
+
+  List<BindableService> getExposedServices() {
+    return ImmutableList.copyOf(exposedServices);
+  }
+}

--- a/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponent.java
+++ b/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponent.java
@@ -47,7 +47,7 @@ public final class PluggableComponent {
   private final List<BindableService> exposedServices = new ArrayList<>();
 
   // A given component API (state store, pubusb, binding etc) can be registered only once.
-  private boolean alreadyAddedStaStore = false;
+  private boolean alreadyAddedStateStore = false;
 
   /**
    * Creates a new pluggable component.
@@ -75,7 +75,7 @@ public final class PluggableComponent {
    * @return The current {@link PluggableComponent} instance, so calls can be chained.
    */
   public PluggableComponent withStateStore(StateStore stateStore) {
-    assert !alreadyAddedStaStore; // No, you cannot add multiple stateStores
+    assert !alreadyAddedStateStore; // No, you cannot add multiple stateStores
 
     exposedServices.add(new StateStoreGrpcComponentWrapper(stateStore));
 
@@ -90,7 +90,7 @@ public final class PluggableComponent {
     // }
 
     // StateStore added and nothing else can be added
-    alreadyAddedStaStore = true;
+    alreadyAddedStateStore = true;
     return this;
   }
 

--- a/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java
+++ b/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java
@@ -207,8 +207,8 @@ public class PluggableComponentServer {
   private void startServers() throws IOException {
     // For each registered server, start it and set up add it to a compound shutdown hook.
     for (var componentNameAndServer : servers.entrySet()) {
-      final var componentName = componentNameAndServer.getKey();
-      final var server = componentNameAndServer.getValue();
+      final String componentName = componentNameAndServer.getKey();
+      final Server server = componentNameAndServer.getValue();
 
       server.start();
       fixUnixDomainSocketFilePermissions(componentName);
@@ -218,9 +218,11 @@ public class PluggableComponentServer {
     Runtime.getRuntime().addShutdownHook(new Thread(() -> {
       // Use stderr here since the logger may have been reset by its JVM shutdown hook.
       System.err.println("*** shutting down gRPC server since JVM is shutting down");
-      servers.forEach((componentName, server) -> {
+      servers.keySet().parallelStream().forEach(componentName -> {
         try {
-          System.err.println("    *** shutting doing server instance " + server.toString());
+          final Server server = servers.get(componentName);
+          System.err.println("    *** shutting down server for component " + componentName
+              + " with instance " + server.toString());
           server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
         } catch (InterruptedException e) {
           e.printStackTrace(System.err);

--- a/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java
+++ b/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java
@@ -40,9 +40,11 @@ import java.util.logging.Logger;
  * Container for pluggable components, responsible for hosting and turning them into a server.
  *
  * <p>Takes care of:
- * 1. setting up Unix Domain Socket files for each pluggable component,
- * 2. environment variable handling,
- * 3. setting up the required server machinery to expose the service and handle requests for it.
+ * <ol>
+ *   <li>setting up Unix Domain Socket files for each pluggable component,</li>
+ *   <li>environment variable handling,</li>
+ *   <li>setting up the required server machinery to expose the service and handle requests for it.</li>
+ * </ol>
  *
  * <p>If you end up needing to extend functionality of this class it might be simpler re-implementing it.
  */

--- a/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java
+++ b/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java
@@ -56,6 +56,8 @@ public class PluggableComponentServer {
    */
   private final String unixDomainSocketFolder;
 
+  private final String unixDomainSocketSuffix;
+
   // Our "list" or registered components. We use a Map to avoid duplicated registrations.
   private final Map<String, Server> servers = new HashMap<>();
 
@@ -64,6 +66,9 @@ public class PluggableComponentServer {
    */
   public PluggableComponentServer() {
     this.unixDomainSocketFolder = getUnixSocketFolderOrAbort();
+    this.unixDomainSocketSuffix = getEnvVarOrDefault(
+        Constants.EnvironmentVariable.DAPR_COMPONENTS_SOCKET_EXTENSION,
+        Constants.Defaults.DAPR_COMPONENTS_SOCKET_EXTENSION);
   }
 
   /**
@@ -190,10 +195,7 @@ public class PluggableComponentServer {
   }
 
   private Path buildPathForComponentUnixDomainSocket(final String componentName) {
-    final String suffix = getEnvVarOrDefault(
-        Constants.EnvironmentVariable.DAPR_COMPONENTS_SOCKET_EXTENSION,
-        Constants.Defaults.DAPR_COMPONENTS_SOCKET_EXTENSION);
-    final String udsFilename = componentName + '.' + suffix;
+    final String udsFilename = componentName + '.' + unixDomainSocketSuffix;
     final Path targetFolder = Paths.get(unixDomainSocketFolder);
     return targetFolder.resolve(udsFilename);
   }

--- a/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java
+++ b/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java
@@ -112,7 +112,7 @@ public class PluggableComponentServer {
     }
 
     // Unix Domain Socket setup
-    final var componentUdsPath = buildPathForComponentUnixDomainSocket(componentName);
+    final Path componentUdsPath = buildPathForComponentUnixDomainSocket(componentName);
     log.info("Configuring server to listen to unix socket domain on file " + componentUdsPath.toAbsolutePath());
     // If file exists, remove it.
     final File componentUdsFile = componentUdsPath.toFile();
@@ -211,6 +211,7 @@ public class PluggableComponentServer {
       final var server = componentNameAndServer.getValue();
 
       server.start();
+      fixUnixDomainSocketFilePermissions(componentName);
       log.info("Started server for " + componentName);
     }
 
@@ -227,6 +228,27 @@ public class PluggableComponentServer {
       });
       System.err.println("*** server shut down");
     }));
+  }
+
+  /*
+   * In other languages/runtimes, such as Go and .Net, we can invoke the syscall `umask` to ensure any UDS this
+   * server creates are world-Read/Writable. It is a simple and elegant way to solve the "permission issue".
+   * This is not the case for Java: the std. SDK doesn't expose such syscall. To address this we perform the
+   * equivalent of `chmod ugo=rwx` on each unix domain socket file we create.
+   */
+  private void fixUnixDomainSocketFilePermissions(final String componentName) {
+    final File unixDomainSocketFile = buildPathForComponentUnixDomainSocket(componentName).toFile();
+    // We assume that the UDS file is created after server.start(). If this assumption proves to be false (for instance,
+    // because this creation happens in another thread) we will be notified.
+    assert unixDomainSocketFile.exists();
+    final boolean notOwnerOnly = false; // Fix permissions for other users too, not only for its owner.
+    // chmod ugo=rwx
+    final boolean success = unixDomainSocketFile.setExecutable(true, notOwnerOnly)
+        && unixDomainSocketFile.setReadable(true, notOwnerOnly)
+        && unixDomainSocketFile.setWritable(true, notOwnerOnly);
+    if (!success) {
+      logFatalAndAbort("Failed to fix permissions for file " + unixDomainSocketFile);
+    }
   }
 
   /**

--- a/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java
+++ b/components-java-sdk/src/main/java/io/dapr/components/server/PluggableComponentServer.java
@@ -1,0 +1,246 @@
+/*
+ * Copyright 2023 The Dapr Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dapr.components.server;
+
+import io.grpc.BindableService;
+import io.grpc.Server;
+import io.grpc.netty.NettyServerBuilder;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.ServerChannel;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.channel.epoll.EpollServerDomainSocketChannel;
+import io.netty.channel.kqueue.KQueue;
+import io.netty.channel.kqueue.KQueueEventLoopGroup;
+import io.netty.channel.kqueue.KQueueServerDomainSocketChannel;
+import io.netty.channel.unix.DomainSocketAddress;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+/**
+ * Container for pluggable components, responsible for hosting and turning them into a server.
+ *
+ * <p>Takes care of:
+ * 1. setting up Unix Domain Socket files for each pluggable component,
+ * 2. environment variable handling,
+ * 3. setting up the required server machinery to expose the service and handle requests for it.
+ *
+ * <p>If you end up needing to extend functionality of this class it might be simpler re-implementing it.
+ */
+public class PluggableComponentServer {
+  private static final Logger log = Logger.getLogger(PluggableComponentServer.class.getName());
+
+  /**
+   * The folder where UDS files will be created.
+   */
+  private final String unixDomainSocketFolder;
+
+  // Our "list" or registered components. We use a Map to avoid duplicated registrations.
+  private final Map<String, Server> servers = new HashMap<>();
+
+  /**
+   * Default Constructor.
+   */
+  public PluggableComponentServer() {
+    this.unixDomainSocketFolder = getUnixSocketFolderOrAbort();
+  }
+
+  /**
+   * Add a new pluggable component to this server.
+   *
+   * @param component the pluggable component to be added.
+   * @return A reference to this {@link PluggableComponentServer}.
+   * @throws IOException if we face problems creating the Unix Domain Socket file for this component.
+   */
+  public PluggableComponentServer registerComponent(final PluggableComponent component) throws IOException {
+    final String componentName = component.getName();
+    if (servers.containsKey(componentName)) {
+      logFatalAndAbort("A pluggable component named '" + componentName
+          + "' was already registered. Aborting.");
+    }
+    final Server componentServer = buildServerForComponent(component);
+
+    this.servers.put(componentName, componentServer);
+
+    return this;
+  }
+
+  /**
+   * Starts all pluggable component servers and wait for them to complete.
+   *
+   * @throws IOException if we face errors starting the servers for each component.
+   */
+  public void run() throws IOException {
+    startServers();
+    blockUntilShutdown();
+  }
+
+  //
+  // Component Registration
+  //
+
+  private Server buildServerForComponent(final PluggableComponent component) throws IOException {
+    final String componentName = component.getName();
+    log.info("Creating server for component " + componentName);
+
+    // Sanity check: we can't do anything with a pluggable component if it isn't exposing any service.
+    final List<BindableService> exposedServices = component.getExposedServices();
+    if (exposedServices.isEmpty()) {
+      logFatalAndAbort("Pluggable component " + componentName
+          + " is not exposing any service or Dapr API. Aborting.");
+    }
+
+    // Unix Domain Socket setup
+    final var componentUdsPath = buildPathForComponentUnixDomainSocket(componentName);
+    log.info("Configuring server to listen to unix socket domain on file " + componentUdsPath.toAbsolutePath());
+    // If file exists, remove it.
+    final File componentUdsFile = componentUdsPath.toFile();
+    if (Files.deleteIfExists(componentUdsPath)) {
+      log.info("Removed previous Unix Socket Descriptor in [" + componentUdsFile + "].");
+    }
+    // Regardless, delete this file on exist. Just good hygiene ;)
+    componentUdsFile.deleteOnExit();
+
+    // Setup the server for handling requests on the UDS
+    final DomainSocketAddress unixSocket = new DomainSocketAddress(componentUdsFile);
+    final EventLoopGroup eventLoopGroup;
+    final Class<? extends ServerChannel> serverChannelClass;
+    if (KQueue.isAvailable()) {
+      log.info("Using KQueue");
+      eventLoopGroup = new KQueueEventLoopGroup();
+      serverChannelClass = KQueueServerDomainSocketChannel.class;
+    } else {
+      log.info("Using Epoll");
+      eventLoopGroup = new EpollEventLoopGroup();
+      serverChannelClass = EpollServerDomainSocketChannel.class;
+    }
+    final NettyServerBuilder serverBuilder = NettyServerBuilder.forAddress(unixSocket)
+        .channelType(serverChannelClass)
+        .workerEventLoopGroup(eventLoopGroup)
+        .bossEventLoopGroup(eventLoopGroup);
+
+    // Add the services exposed by this pluggable component
+    exposedServices.forEach(serverBuilder::addService);
+
+    return serverBuilder.build();
+  }
+
+  //
+  // Unix Domain Socket - path handling, creation and auxiliary methods
+  //
+
+  private static String getUnixSocketFolderOrAbort() {
+    // Environment variable "parsing"
+    final String unixSocketPath = getEnvVarOrDefault(
+        Constants.EnvironmentVariable.DAPR_COMPONENTS_SOCKETS_PATH,
+        Constants.Defaults.DAPR_COMPONENTS_SOCKETS_PATH);
+    // Validate directory exists and is a writable folder
+    final File targetFolder = new File(unixSocketPath);
+    final boolean isValidPath = targetFolder.exists()
+        && targetFolder.isDirectory()
+        && targetFolder.canWrite();
+    if (!isValidPath) {
+      final String errMsg = "ERROR: path '" + unixSocketPath + "' is not an existing writable directory. Aborting";
+      logFatalAndAbort(errMsg);
+    }
+
+    log.info("Unix Domain Socket files will be created on " + targetFolder.toPath().toAbsolutePath());
+
+    return unixSocketPath;
+  }
+
+  // While this does what its name says, it is not really testable.
+  // We should consider if migrating this to exceptions would be better.
+  // Created components-java-sdk#21 to track it.
+  private static void logFatalAndAbort(String errMsg) {
+    log.severe(errMsg);
+    System.err.println(errMsg);
+    System.exit(1);
+  }
+
+  private static String getEnvVarOrDefault(final String envVarName, final String defaultValue) {
+    String value = System.getenv(envVarName);
+    if (value == null) {
+      log.info("Environment variable " + envVarName + " is not defined. "
+          + " Assuming default value of " + defaultValue);
+      return defaultValue;
+    }
+    return value;
+  }
+
+  private Path buildPathForComponentUnixDomainSocket(final String componentName) {
+    final String suffix = getEnvVarOrDefault(
+        Constants.EnvironmentVariable.DAPR_COMPONENTS_SOCKET_EXTENSION,
+        Constants.Defaults.DAPR_COMPONENTS_SOCKET_EXTENSION);
+    final String udsFilename = componentName + '.' + suffix;
+    final Path targetFolder = Paths.get(unixDomainSocketFolder);
+    return targetFolder.resolve(udsFilename);
+  }
+
+
+  //
+  // Server lifecycle management methods
+  //
+
+
+  private void startServers() throws IOException {
+    // For each registered server, start it and set up add it to a compound shutdown hook.
+    for (var componentNameAndServer : servers.entrySet()) {
+      final var componentName = componentNameAndServer.getKey();
+      final var server = componentNameAndServer.getValue();
+
+      server.start();
+      log.info("Started server for " + componentName);
+    }
+
+    Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+      // Use stderr here since the logger may have been reset by its JVM shutdown hook.
+      System.err.println("*** shutting down gRPC server since JVM is shutting down");
+      servers.forEach((componentName, server) -> {
+        try {
+          System.err.println("    *** shutting doing server instance " + server.toString());
+          server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+          e.printStackTrace(System.err);
+        }
+      });
+      System.err.println("*** server shut down");
+    }));
+  }
+
+  /**
+   * Await termination on the main thread since the grpc library uses daemon threads.
+   */
+  private void blockUntilShutdown() {
+    // There probably is a more elegant way of doing this but, for the time being, this should do.
+    servers.values()
+        .parallelStream()
+        .map(server -> {
+          try {
+            server.awaitTermination();
+            return true;
+          } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+          }
+        });
+  }
+}

--- a/pom.xml
+++ b/pom.xml
@@ -24,6 +24,7 @@
         <grpc.version>1.42.1</grpc.version>
         <protobuf.version>3.17.3</protobuf.version>
         <springboot.version>2.3.5.RELEASE</springboot.version>
+        <netty-transport-native.version>4.1.79.Final</netty-transport-native.version>
 
         <spotbugs.version>4.7.3.0</spotbugs.version>
         <spotbugs.fail>true</spotbugs.fail>

--- a/pom.xml
+++ b/pom.xml
@@ -15,6 +15,7 @@
     <modules>
         <module>components-java-sdk-autogen</module>
         <module>components-java-sdk</module>
+        <module>components-java-sdk-examples</module>
     </modules>
 
     <properties>

